hub/core/storage/ipfs.py

```diff
@@ -0,0 +1,483 @@
+from aifc import Error
+from distutils.log import error
+import hub
+import time
+import json
+import logging
+import dag_cbor
+import requests
+import pandas as pd
+
+from io import StringIO,BytesIO
+from requests.exceptions import HTTPError
+from typing import Optional, Set, Sequence
+from hub.core.storage.provider import StorageProvider
+from typing import Union, List
+
+from ipfshttpclient.multipart import stream_files, stream_directory
+
+
+logger = logging.getLogger("ipfsspec")
+
+class IPFSProvider(StorageProvider):
+    def __init__(
+        self,
+        coreurl:str=None, # Core URL to use
+        cid:str='',
+        storage_type:str=None, # specify type of gateway (e.g. Infura, Estuary, Web3.Storage, local node...)
+        api_key:str=None, # if applicable, api key for access to storage service
+        fpath:str=None,
+    ) -> None:
+        """Initialize the object, assign credentials if required."""
+        super().__init__()
+        self.coreurl = coreurl if coreurl is not None else 'https://ipfs.infura.io:5001/api/v0'
+        self.cid = cid
+        self.gw = IPFSGateway(url=self.coreurl)
+        self.storage_type = storage_type
+        self.api_key = api_key
+        self.cids = None
+        self.fpath = fpath
+        self.stored = False
+
+    def __getitem__(self, path, **kwargs):
+        """Gets the object present at the path."""
+        if self.cid != '':
+            try:
+                self.links = self._get_links(self.cid)
+                self.ordered_links = self.get_hash(self.links, {})
+                cid = self.ordered_links[path]
+                res, content = self.gw.cat(cid)
+                b = bytes(content,'utf-8')
+                if res.status_code == 200:                
+                    return b
+                else:
+                    raise HTTPError (parse_error_message(res))
+            except StopIteration:
+                print(f"Path requested: {path}")
+            except FileNotFoundError:
+                print(f"Path requested: {path}, filenotfound")
+                raise KeyError(path)
+            except AssertionError:
+                print(f'Status code: {res.status_code}, Assertion error, content is {content}, path is {path}, cid query is {cid}, query is {query}, and in bytes we have {b}')
+            except TypeError:
+                print('Got type error.')
+        else:
+            raise KeyError(path)
+
+    def __setitem__(self, path, value):
+        """Sets the object present at the path with the value"""
+        if not self.stored:
```

Comment:
why. are we keeping this flag? will only the first __setitem__ work? after which it will just return True instead of actually doing setitem?